File: algo-balanced-parentheses/js/balanceParens.js
balanceParens = (str) => {
    const noKeep = new Set();
    const parenIndexStack = [];
    let output = '';
  
    for (let i = 0; i < str.length; i++) {
      if (str[i] == '(') {
        parenIndexStack.push(i);
      } else if (str[i] == ')') {
        if (parenIndexStack.length == 0) {
          noKeep.add(i);
        } else {
          parenIndexStack.pop();
        }
      }
    }

    parenIndexStack.forEach(i=> noKeep.add(i))
    str.split('').forEach((e, i)=> {if (!noKeep.has(i)) output += e})
    return output;
}

module.exports = { balanceParens }

File: algo-balanced-parentheses/python/balance_parens.py
def balance_parens(str):
    noKeep = set()
    parenIndexStack = []
    for i in range(len(str)):
        if str[i] == '(':
            parenIndexStack.append(i)
        elif str[i] == ')':
            if not parenIndexStack:
                noKeep.add(i)
            else:
                parenIndexStack.pop()

    for i in parenIndexStack:
        noKeep.add(i)

    return ''.join(str[i] for i in range(len(str)) if i not in noKeep)
